### PR TITLE
Extraction: get rid of Big module

### PR DESCRIPTION
**Pull request by liyishuai**
**Kind:** bug fix.

- [x] Entry added in CHANGES.

### CODE REVIEW

**Comment by gares** [comment]
I don't get what is the purpose of this patch.
The change links Coq (the extraction plugin, precisely) against this big.cmx file (I don't know where it comes from actually).
If Coq was running before this change, then it does not need this module.

Can you explain what are you trying to achieve?

**Comment by liyishuai** [comment]
As in QuickChick/QuickChick#102, I tried to extract `Z` to `Big.big_int`, but could not compile the extracted OCaml program.

**Comment by liyishuai** [comment]
Are there existing programs that use `ExtrOcamlZBigInt`? I have a wild guess that this module is not tested.

**Comment by gares** [comment]
I don't know. But my guess is that `big.cmx` is installed by Coq so that you can link against it the extracted code (as in `ocamlopt extracted.ml -I /wherever_coq/lib/ big.cmx`). Linking the .cmx inside Coq makes little sense to me.

**Comment by liyishuai** [comment]
`big.cmx` was not installed before this PR.
This patch puts `big.cmx` under `lib/coq/plugins/extraction`.

**Comment by liyishuai** [comment]
Update: I'm now proposing to make `Big` a module in `coq.lib` package, so that users can really use `ExtrOcamlZBigInt`.

**Comment by liyishuai** [comment]
The test suite now requires Coq to be installed. Shall we update the CI script or find other ways around?

**Comment by Zimmi48** [comment]
The test suite already requires Coq to be installed unless the option `-local` is passed (in which case, "installation" is not possible).

**Comment by liyishuai** [comment]
The failing tests on [Travis](https://travis-ci.org/coq/coq/jobs/416629704#L86), [AppVeyor](https://ci.appveyor.com/project/coq/coq/build/master~6383?fullLog=true#L243) are both configured with `-local`.
The [GitLab](https://gitlab.com/coq/coq/-/jobs/89660622) test installs with prefix `_install_ci`, in which the packages might be invisible from OCamlfind.

**Comment by Zimmi48** [comment]
The test-suite failures are indeed:
```
ocamlfind: Package `coq.lib' not found
```
However, I don't see why the location of the installation should change whether `ocamlfind` find this package or not.
Note that the Nix test fails even earlier (during compilation) with
```
OCAMLC    lib/big.mli
File "lib/big.mli", line 11, characters 15-30:
Error: Unbound module Big_int
```
I don't know how to interpret this. The dependency bignums is available and can be found by the rest of the system. I'll let make experts comment further.

**Comment by SkySkimmer** [comment]
We don't pass -package num when compiling lib so it only works when num is installed with the standard ocaml files.

**Comment by SkySkimmer** [comment]
I don't think it makes sense to link extracted programs with coq.lib. This is stuff like futures, remotecounter, our warning system...
We could instead have a new independent (maybe depending on clib but not lib) package coq.extractionlib (which for now would only contain this Big module).

**Review by maximedenes — DISMISSED** [review]
I agree with @SkySkimmer.

**Comment by liyishuai** [comment]
Changes in process.
Could someone instruct me how to include the new `/extractionlib` directory in the building and installation process, or push to my branch if it's more instant?

**Comment by liyishuai** [comment]
I've figured out the installation process. Now extraction-related test suite works with Coq installed under my OPAM directory.

File: META.coq.in
Comment (gares) [inline META.coq.in]: Is this dependency true?

**Review by gares — CHANGES_REQUESTED** [review]
fix the dependency of the META file

**Comment by Zimmi48** [comment]
@liyishuai Note that @coqbot requested a rebase because it won't run CI unless the PR is up-to-date with the master branch.

**Review by ejgallego — COMMENTED** [review]
First, why is extraction using its own big num implementation instead of a more standard one [either zarith or num] ?

Second, this library should be private to extraction, so please place it in `plugins/extraction/runtime`. We will take more advantage of this if we ever get a modular build system.

**Review by gares — APPROVED** [review]
I agree with @ejgallego that the extractionlib directory should be inside the extraction folder and not at the toplevel.

**Comment by liyishuai** [comment]
Here's my proposed solution: replace `Big.*` in [ExtrOcamlZBigInt](https://coq.inria.fr/library/Coq.extraction.ExtrOcamlZBigInt.html) with their `Big_int.*` definition in `big.ml`, so that the extracted OCaml code only rely on [num](https://opam.ocaml.org/packages/num/).

Maybe the author of fb6ae3d50279005f75deb273de1d0067a5fa089a has better suggestions.

**Comment by ejgallego** [comment]
> Here's my proposed solution: replace Big.* in ExtrOcamlZBigInt with their Big_int.* definition in big.ml, so that the extracted OCaml code only rely on num.
> 
> Maybe the author of fb6ae3d has better suggestions.

That looks fine to me, however note that `num` is deprecated, I suggest you try to target `zarith`.

**Comment by liyishuai** [comment]
I don't quite understand the [error](https://ci.appveyor.com/project/coq/coq/build/master~7020#L231) with AppVeyor.
Should the [GitLab](https://gitlab.com/coq/coq/pipelines/30171110) CI install `zarith` in script or in container?

File: META.coq.in
Comment (ejgallego) [inline META.coq.in]: I am not sure this is accurate, right? As far as I understand the plugin itself doesn't require `zarith` to run, right?

**Comment by ejgallego** [comment]
> I don't quite understand the [error](https://ci.appveyor.com/project/coq/coq/build/master%7E7020#L231) with AppVeyor.

You need to install the cygwin `gmp` library. This is a C library and has to be installed outside of OPAM I think. We'd need an AppVeyor / Cywin expert here.

> Should the [GitLab](https://gitlab.com/coq/coq/pipelines/30171110) CI install `zarith` in script or in container?

Indeed, see `dev/ci/README.ci` for instructions.

**Review by ejgallego — COMMENTED** [review]
Given the status of OCaml's arithmetic libraries it seems very reasonable to me to support `zarith`. In particular we lack the infrastructure to properly install plugin-runtime ML libraries [for now] thus I do highly approve of this PR.

Note however that test-suite, and a few couple of nits have still to be fixed.

File: META.coq.in
Comment (liyishuai) [reply META.coq.in]: The plugin doesn't require `zarith` to compile Coq or extract it to OCaml. Only when using `Big_int_Z` do we need `zarith` to compile the extracted OCaml code, but my [implementation](https://github.com/coq/coq/pull/8252/files#diff-a0236bab39290033da13a82f61a4dd0dR726) always requires `zarith`.
Is there a way to auto-detect usage of modules, in compilation stage?

Comment (ejgallego) [reply META.coq.in]: You can provide a `coq.plugins.extraction.runtime` package, but it should not depend on `coq.ltac` for example. This meta file is indeed for compiling the plugin itself so `zarith` should not be there.

**Comment by Zimmi48** [comment]
I bet https://github.com/coq/coq/blob/f1482433ff225831d9937753f946cff2577b9309/dev/ci/appveyor.bat#L34 is the line you need to change to fix AppVeyor.
Maybe the Windows build scripts would need to be updated as well (not sure). I'll let @MSoegtropIMC decide.

**Comment by MSoegtropIMC** [comment]
For the gitlab windows build (which is the one which produces the "official" installer), one would have to:

- add a make shell function for https://github.com/ocaml/Zarith/archive/release-1.7.tar.gz in makecoq_mingw.sh
- call this function in mak_coq in makecoq_mingw.sh
- add the corresponding cygwin module in MakeCoq_MinGW.bat
- possibly manually install the created ocaml libraries where they belong in makecoq_mingw.sh

I could also do this after the integration of this PR in a separate PR.

One more note: opam on cygwin has a module which automatically installs cygwin modules if they are required by an opam module. One just has to install and initialize this module first. If this is of interest, I can dig out the details.

**Comment by Zimmi48** [comment]
>One more note: opam on cygwin has a module which automatically installs cygwin modules if they are required by an opam module. One just has to install and initialize this module first. If this is of interest, I can dig out the details.

That sounds amazing.

File: META.coq.in
Comment (ejgallego) [reply META.coq.in]: The runtime package should contain the minimal set of dependencies that _extracted_ code need to run.

File: plugins/extraction/ExtrOcamlZBigInt.v
Comment (liyishuai) [inline plugins/extraction/ExtrOcamlZBigInt.v]: One thing to notice is that: I tried to use the [Z](https://github.com/ocaml/Zarith/blob/master/z.mlip) module for simplicity, but the extracted code had name conflict in it, even if I blacklisted `Z` (which only renames the file?). We probably need module-level blacklist mechanisms.

**Comment by jhjourdan** [comment]
Why using the `Big_int_Z` compatibility layer instead of the actual interface of the ZArith library ?

**Comment by liyishuai** [comment]
Then the extracted code will use OCaml `Z` in the definition of Coq `Z`, causing a conflict.

**Comment by liyishuai** [comment]
(Since this PR is mentioned in Coq Club) Are there still changes required on my side? Is someone handling the CI issue?

**Comment by gares** [comment]
I think you should rebase to resolve conflicts. The issue about bedrock may be spurious, the one about nix is more likely to require a fix.

File: plugins/extraction/extract_env.ml
Comment (liyishuai) [inline plugins/extraction/extract_env.ml]: Should this be `zarith` or `coq.extraction.runtime`?

Comment (ejgallego) [reply plugins/extraction/extract_env.ml]: I think for now `zarith` should be fine. The whole point of using `zarith` is not to provide a special runtime for extraction. Also, Coq may require zarith soon so that could be fine. Now, it could fail if the user doesn't have it.

**Comment by maximedenes** [comment]
@coq/ci-maintainers can you suggest what approach we should follow here?

**Comment by ejgallego** [comment]
> @coq/ci-maintainers can you suggest what approach we should follow here?

This is more of a build maintainers team issue.

Dune can support this quite easily I think, but until we are migrated I am afraid I cannot help a lot more.

**Comment by SkySkimmer** [comment]
What's the issue? I'm getting lost in the discussion.

**Comment by ejgallego** [comment]
> What's the issue? I'm getting lost in the discussion.

The issue is about the runtime dependencies of extracted code.

**Comment by liyishuai** [comment]
I'm rebasing this branch. Which sub-directory should I put the new changelog in?

**Comment by Zimmi48** [comment]
I suppose either `11-infrastructure-and-dependencies` or `12-misc`.

**Review by Zimmi48 — COMMENTED** [review]
The `zarith` dependency is not documented anywhere. I think it should be.

**Comment by liyishuai** [comment]
Where should I put `zarith` documentations, Changelog?

**Comment by Zimmi48** [comment]
Yes, indeed. Changelog at least. Maybe in the stdlib doc too (i.e. coqdoc comments).

**Comment by liyishuai** [comment]
Dependency on `zarith` documented in Changelog and CoqDoc.

**Review by Zimmi48 — APPROVED** [review]

**Comment by Zimmi48** [comment]
@liyishuai It seems that the test-suite is broken.

**Comment by liyishuai** [comment]
The complaint is:
> ocamlfind: Package `zarith' not found

I need help with installing `zarith` for test suite. Maintainers can push fixes to my branch directly if that's more instant.

**Comment by Zimmi48** [comment]
Likely you need to install it with opam on Azure (macOS + Windows), in the Docker image (see `dev/ci/docker`), and in the `default.nix` file (cc @coq/ci-maintainers @coq/nix-maintainers).

**Comment by ejgallego** [comment]
@liyishuai that's not a trivial task due to the 32bit buld, see #8743 tho where the install is done.

**Review by maximedenes — COMMENTED** [review]
This PR seems to break the test suite.

**Comment by liyishuai** [comment]
The changes here require ZArith installed, probably depends on #8743. Changing to draft for now.

**Comment by ejgallego** [comment]
@liyishuai do you mind I rebase this on top of #11742  ?

**Comment by liyishuai** [comment]
> @liyishuai do you mind I rebase this on top of #11742 ?

No problem, please proceed.

**Comment by liyishuai** [comment]
Are we keeping the unused `big.ml`?

**Comment by ejgallego** [comment]
> Are we keeping the unused `big.ml`?

Oh indeed I see; we should indeed just drop it, together with the META changes then, right?

**Comment by liyishuai** [comment]
Right, the extracted code only require `zarith`.

**Comment by ejgallego** [comment]
> Right, the extracted code only require `zarith`.

See my latest push / adjust as you wish, releasing the lock.

**Comment by liyishuai** [comment]
Depends on coq/bignums#44

**Comment by liyishuai** [comment]
Ready to merge (?)

File: doc/changelog/12-misc/08252-extract_big.rst
Comment (gares) [inline doc/changelog/12-misc/08252-extract_big.rst:3]: I'm a bit confused by the changelog entry. Would you mind making it more clear what `zarith` is (to me is also a Coq stdlib module). The same for big_int. Maybe a link to these libraries?

Comment (gares) [reply doc/changelog/12-misc/08252-extract_big.rst]: thanks

**Comment by liyishuai** [comment]
Are there more TODOs for me before merging?

**Comment by gares** [comment]
I believe someone from @coq/extraction-maintainers should take care of this PR

**Comment by maximedenes** [comment]
Will do, sorry for the delay.

File: theories/extraction/ExtrOcamlNatBigInt.v
Comment (gares) [inline theories/extraction/ExtrOcamlNatBigInt.v:31]: is this valid ocaml?

**Review by gares — DISMISSED** [review]
I think we should have a 3 test suite files, one per ExtrOcaml*, that uses each and every Extract Constant, and we should at least check the obtained file compiles.

**Comment by liyishuai** [comment]
The test suite for [SimpleIO](https://github.com/Lysxia/coq-simple-io/blob/master/test/TestExtraction.v) includes one term per module, so that the whole module gets extracted. Should we make a similar test here?

File: theories/extraction/ExtrOcamlNatBigInt.v
Comment (liyishuai) [reply theories/extraction/ExtrOcamlNatBigInt.v]: Now fixed and included in test suite.

File: theories/extraction/ExtrOcamlBigIntConv.v
Comment (silene) [inline theories/extraction/ExtrOcamlBigIntConv.v:70]: You need to parenthesize the code, otherwise this will not compile.

File: theories/extraction/ExtrOcamlBigIntConv.v
Comment (silene) [inline theories/extraction/ExtrOcamlBigIntConv.v:83]: Same here.

That said, I do not even understand what this file is about. It adds tens of axioms. This does not make sense to me.

File: theories/extraction/ExtrOcamlNatBigInt.v
Comment (silene) [inline theories/extraction/ExtrOcamlNatBigInt.v:43]: Spaces are copied verbatim into the output, so you should refrain from indenting the Coq file, as it makes the resulting OCaml file unreadable.

Also, this should really be part of the testsuite, because there seems to be some missing parenthesis here.

**Review by silene — CHANGES_REQUESTED** [review]

File: theories/extraction/ExtrOcamlBigIntConv.v
Comment (liyishuai) [reply theories/extraction/ExtrOcamlBigIntConv.v:70]: Do we need to parenthesize code that are not inline-extracted?

File: theories/extraction/ExtrOcamlBigIntConv.v
Comment (liyishuai) [reply theories/extraction/ExtrOcamlBigIntConv.v:83]: I don't quite understand this module either. Could remove this instead.

File: theories/extraction/ExtrOcamlBigIntConv.v
Comment (silene) [reply theories/extraction/ExtrOcamlBigIntConv.v:70]: Good point. That said, it is hard to predict whether some code will be inlined, as the user can always type `Extraction Inline foo.` after the fact (or the extraction mechanism might have decided to inline a symbol for heuristic reasons).

Comment (liyishuai) [reply theories/extraction/ExtrOcamlBigIntConv.v:70]: It should be an improvement for Coq that handles parentheses in extracted code..

File: theories/extraction/ExtrOcamlZBigInt.v
Comment (Lysxia) [inline theories/extraction/ExtrOcamlZBigInt.v:35]: Use `Z` rather than `Big_int_Z`. The latter is meant to be a drop-in replacement for `Big_int` which we haven't been using in the first place.

Comment (liyishuai) [reply theories/extraction/ExtrOcamlZBigInt.v:35]: My reply three years ago: https://github.com/coq/coq/pull/8252#issuecomment-428233520
Not sure if the problem has been solved or not, will try it though.

**Comment by liyishuai** [comment]
Removed `ExtrOcamlBigIntConv` per https://github.com/coq/coq/pull/8252#discussion_r639706154.
Propose merging upon CI passed, and address https://github.com/coq/coq/pull/8252#discussion_r709298107 in a separate PR.

**Comment by liyishuai** [comment]
SimpleIO depends on cppo >= 1.6.8, but the CI installs 1.6.7. Did I miss some configuration?

**Comment by silene** [comment]
> SimpleIO depends on cppo >= 1.6.8, but the CI installs 1.6.7. Did I miss some configuration?

The CI is independent from the Opam package. You get to use the version of `cppo` that is embedded in the Docker image. Presumably, the Docker image was built before `cppo` 1.6.8 was released.

**Review by Alizter — APPROVED** [review]

**Comment by liyishuai** [comment]
Now I understand. Mistook `Extraction Inline` for `Extract Inlined Constant` previously. Will add the parentheses later today.

**Review by silene — APPROVED** [review]

**Comment by Alizter** [comment]
@liyishuai Just to be sure, have you addressed @gares' comments? Since I see a test in the test-suite I assume so. Also are @maximedenes / @SkySkimmer's comments still relevant?

**Comment by Alizter** [comment]
Alright will merge soon

**Comment by liyishuai** [comment]
I've tested QuickChick/QuickChick#102, which depends on and works fine with the changes here.

The CI doesn't capture runtime bugs in QuickChick until #15055.

**Pull request merged**
